File: ordenacoes/quicksort.py
# Estrutura de Dados II
# Matheus Marinho
# Quick Sort

def particaoVetor(left, right, i, j, arr):
    i = left
    j = right
    x = arr[(i+j)//2] # Escolha do pivô sendo o elemento central

    while True:
        # Buscando elementos maiores que o pivô (>)
        while arr[i] < x:
            i += 1 
        # Buscando elementos menores que o pivô (<)
        while arr[j] > x:
            j -= 1
        # Troca, se necessário:
        if i <= j:
            aux = arr[i]
            arr[i] = arr[j]
            arr[j] = aux
            i += 1
            j -= 1
        else:
            break 

    return i, j

def ordenaVetor(left, right, arr):
    if left < right:
        i, j = left, right
        i, j = particaoVetor(left, right, i, j, arr)
        ordenaVetor(left, j, arr) # Ordenaçao (<)
        ordenaVetor(i, right, arr) # Ordenacao (>)

def quickSort(arr):
    ordenaVetor(0, len(arr)-1, arr)
    return arr

# Input:    
arr = [50, 16, 23, 5, 7, 12, 10]

# Output:
# Array original:  [50, 16, 23, 5, 7, 12, 10]
# Array ordenado:  [5, 7, 10, 12, 16, 23, 50]

print("Array original: ", arr)
print("Array ordenado: ", quickSort(arr))
print("Menor valor: ", arr[0])
print("Maior valor: ", arr[len(arr)-1])

File: ordenacoes/quicksort.c
/* Estrutura de Dados II
    Matheus Marinho   
    Quick Sort
*/

#include <stdio.h>
#include <stdlib.h>

// Quick Sort por partes funcional

/* Partição do vetor para a ordenação */

void particaoVetor(int esq, int dir, int *i, int *j, int arr[]){
    /* Particionando o vetor de entrada */
    *i = esq;
    *j = dir;
    int aux;
    int x = arr[(*i + *j)/2];  /* Escolha de um pivô baseado na média dos valores */

    do{
        while(arr[*i] < x) (*i)++;  /* Buscamos um elemento maior que o pivô (>) */
        while(arr[*j] > x) (*j)--;  /* Buscamos um elemento menor que o pivô (<) */

        if (*i <= *j) { /* Troca dos elementos */
            aux = arr[*i];
            arr[*i] = arr[*j];
            arr[*j] = aux;
            (*i)++;
            (*j)--;
        }
    }while(*i <= *j);
}

/* Processo de ordenação do quickSort */
void ordenaVetor(int esq, int dir, int arr[]){
    /* Particione o vetor */
    int i, j;
    particaoVetor(esq, dir, &i, &j, arr);
   
    /* Já tendo o vetor particionado, organiza-se o vetor por recursão */ 
   
    if(esq < j){
        ordenaVetor(esq, j, arr);
    }
    if(i < dir){
        ordenaVetor(i, dir, arr);
    }
}

void quickSort(int arr[], int tam){
    ordenaVetor(0, tam-1, arr);
}

// Array original:  [50, 16, 23, 5, 7, 12, 10]

int main(){
    int arr[] = {50, 16, 23, 5, 7, 12, 10};
    int size = sizeof(arr)/sizeof(arr[0]);
    
    printf("Vetor original: ");
    for(int i = 0; i < size; i++){
        printf("%d ", arr[i]);
    }
    printf("\n");
    printf("\n");
    
    // Chamando a função para ordenar o array
    quickSort(arr, size); 
    
    printf("Vetor ordenado: ");
    for(int i = 0; i < size; i++){
        printf("%d ", arr[i]);
    }
    printf("\n");
    printf("\n");
    printf("Menor valor do array: [%d]", arr[0]);
    printf("\n");
    printf("Maior valor do array: [%d]", arr[size-1]);
    printf("\n");

    return 0;
}